### PR TITLE
Vec 190 tls

### PR DESCRIPTION
**Pull request by jdogmcsteezy**
Add tls and auth support. Auth support is added via the token_manager.go. It inserts the token into each "Unary" request behind the scenes and has its own tend thread to manage to refresh the token. In the initial connection to seeds we block all other requests from authenticating until at least one is finished. 

Again, there are no tests here but the asvec repo has tests for TLS, mTLS, and TLS + Auth (auth requires tls). 

Lastly, this PR is being compared against [VEC-155-admin-client](https://github.com/aerospike/avs-client-go/tree/VEC-155-admin-client) instead of main to display a cleaner diff. After it is review I will merge it into main

### CODE REVIEW

File: token_manager.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+	tm.logger.Debug("scheduling token refresh")
+
+	tm.refreshScheduled = true
+	timer := time.NewTimer(0)
+
+	go func() {
+		for {
+			conn, err := getConn()
+			if err != nil {
+				tm.logger.WarnContext(context.Background(), "failed to refresh token", slog.Any("error", err))
```
Comment (reugn) [inline token_manager.go:154]: Use `logger.Warn` instead.

File: token_manager.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+
+			_, err = tm.RefreshToken(ctx, conn)
+			if err != nil {
+				tm.logger.WarnContext(context.Background(), "failed to refresh token", slog.Any("error", err))
```
Comment (reugn) [inline token_manager.go:161]: Use `logger.Warn` instead.

File: token_manager.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+	return true, nil
+}
+
+// ScheduleRefresh schedules the token refresh using the provided function to
+// get the gRPC client connection. This is not threadsafe. It should only be
+// called once.
+func (tm *tokenManager) ScheduleRefresh(getConn func() (*grpc.ClientConn, error)) {
+	if tm.refreshScheduled {
+		tm.logger.WarnContext(context.Background(), "refresh already scheduled")
```
Comment (reugn) [inline token_manager.go:142]: Use `logger.Warn` instead.

File: token_manager.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	claims := strings.Split(resp.GetToken(), ".")
+	decClaims, err := base64.RawURLEncoding.DecodeString(claims[1])
+
+	if err != nil {
+		return false, fmt.Errorf("%s: %w", "failed to authenticate", err)
+	}
+
+	tokenMap := make(map[string]interface{}, 8)
```
Comment (reugn) [inline token_manager.go:100]: interface{} -> any

File: token_manager.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		return true
+	}
+
+	return time.Now().After(expiredTime.(time.Time))
+}
+
+// RefreshToken refreshes the authentication token using the provided gRPC client connection.
+// It returns a boolean indicating if the token was successfully refreshed and
+// an error if any. It is not thread safe.
+func (tm *tokenManager) RefreshToken(ctx context.Context, conn grpc.ClientConnInterface) (bool, error) {
```
Comment (reugn) [inline token_manager.go:75]: Is there any advantage to returning (bool, error) over error?

File: channel_provider.go [hunk 6]
```diff
@@ -10,71 +12,113 @@ import (
...
-	}
+		msg := "seeds cannot be nil or empty"
+		logger.Error(msg)
+		return nil, errors.New(msg)
+	}
+
+	// Create a token manager if username and password are provided.
+	var token *tokenManager
+	if username != nil || password != nil {
+		if username == nil || password == nil {
```
[the remark this answers is not in the document]
Comment (jdogmcsteezy) [reply channel_provider.go:78]: I think so. It checks that both username and password are provided. Do you have a suggestion for something cleaner?

File: token_manager.go (hunk 5 above)
Comment (jdogmcsteezy) [reply token_manager.go:75]: It was mostly an artifact of an older design when I was going to make TokenManager threadsafe. I still thought I needed it but now I do not :). Thanks!

File: channel_provider.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (reugn) [reply channel_provider.go:78]: The outer if is redundant.

File: token_manager.go (hunk 5 above)
Comment (reugn) [reply token_manager.go:75]: If there are no plans for this boolean, consider not returning it.

File: channel_provider.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdogmcsteezy) [reply channel_provider.go:78]: If username is nil and password is not nil I want to generate an error. If password is nil and username is not nil I also want to generate an error

File: token_manager.go (hunk 5 above)
Comment (jdogmcsteezy) [reply token_manager.go:75]: already made this change. Just waiting to submit

**Review by reugn — APPROVED** [review]

**Comment by jdogmcsteezy** [comment]
Opps need to merge this branch into main, not VEC-155-admin-client

**Pull request merged**